### ISSUE TITLE
Handle Byte Order Mark while importing Constraints file.

### ISSUE THREAD

**Issue by aakashysharma** [webanno/webanno#95]
After exporting a constraint rule file and attempting to upload again to project, sometimes encounter into this error.

Stacktrace

```

de.tudarmstadt.ukp.clarin.webanno.constraints.grammar.TokenMgrError: Lexical error at line 1, column 1.  Encountered: "\ufffd" (65533), after : ""
     at de.tudarmstadt.ukp.clarin.webanno.constraints.grammar.ConstraintsGrammarTokenManager.getNextToken(ConstraintsGrammarTokenManager.java:643)
     at de.tudarmstadt.ukp.clarin.webanno.constraints.grammar.ConstraintsGrammar.jj_scan_token(ConstraintsGrammar.java:637)
     at de.tudarmstadt.ukp.clarin.webanno.constraints.grammar.ConstraintsGrammar.jj_3R_10(ConstraintsGrammar.java:409)
     at de.tudarmstadt.ukp.clarin.webanno.constraints.grammar.ConstraintsGrammar.jj_3_4(ConstraintsGrammar.java:386)
     at de.tudarmstadt.ukp.clarin.webanno.constraints.grammar.ConstraintsGrammar.jj_3R_7(ConstraintsGrammar.java:399)
     at de.tudarmstadt.ukp.clarin.webanno.constraints.grammar.ConstraintsGrammar.jj_3_1(ConstraintsGrammar.java:392)
     at de.tudarmstadt.ukp.clarin.webanno.constraints.grammar.ConstraintsGrammar.jj_2_1(ConstraintsGrammar.java:279)
     at de.tudarmstadt.ukp.clarin.webanno.constraints.grammar.ConstraintsGrammar.Parse(ConstraintsGrammar.java:17)
     at de.tudarmstadt.ukp.clarin.webanno.project.page.ProjectConstraintsPanel$ImportForm.importAction(ProjectConstraintsPanel.java:305)
     at de.tudarmstadt.ukp.clarin.webanno.project.page.ProjectConstraintsPanel$ImportForm.access$400(ProjectConstraintsPanel.java:261)
     at de.tudarmstadt.ukp.clarin.webanno.project.page.ProjectConstraintsPanel$ImportForm$1.onSubmit(ProjectConstraintsPanel.java:280)
     at org.apache.wicket.markup.html.form.Form.delegateSubmit(Form.java:1305)
     at org.apache.wicket.markup.html.form.Form.process(Form.java:966)
     at org.apache.wicket.markup.html.form.Form.onFormSubmitted(Form.java:788)
     at org.apache.wicket.markup.html.form.Form.onFormSubmitted(Form.java:721)
     at java.lang.reflect.Method.invoke(Method.java:497)
     at org.apache.wicket.RequestListenerInterface.internalInvoke(RequestListenerInterface.java:258)
     at org.apache.wicket.RequestListenerInterface.invoke(RequestListenerInterface.java:216)
     at org.apache.wicket.core.request.handler.ListenerInterfaceRequestHandler.invokeListener(ListenerInterfaceRequestHandler.java:243)
     at org.apache.wicket.core.request.handler.ListenerInterfaceRequestHandler.respond(ListenerInterfaceRequestHandler.java:236)
     at org.apache.wicket.request.cycle.RequestCycle$HandlerExecutor.respond(RequestCycle.java:890)
     at org.apache.wicket.request.RequestHandlerStack.execute(RequestHandlerStack.java:64)
     at org.apache.wicket.request.cycle.RequestCycle.execute(RequestCycle.java:261)
     at org.apache.wicket.request.cycle.RequestCycle.processRequest(RequestCycle.java:218)
     at org.apache.wicket.request.cycle.RequestCycle.processRequestAndDetach(RequestCycle.java:289)
     at org.apache.wicket.protocol.http.WicketFilter.processRequestCycle(WicketFilter.java:259)
     at org.apache.wicket.protocol.http.WicketFilter.processRequest(WicketFilter.java:201)
     at org.apache.wicket.protocol.http.WicketFilter.doFilter(WicketFilter.java:282)
     at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:235)
     at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:206)
     at org.springframework.orm.jpa.support.OpenEntityManagerInViewFilter.doFilterInternal(OpenEntityManagerInViewFilter.java:147)
     at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:76)
     at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:235)
     at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:206)
     at net.bull.javamelody.MonitoringFilter.doFilter(MonitoringFilter.java:164)
     at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:235)
     at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:206)
     at de.tudarmstadt.ukp.clarin.webanno.webapp.WebAnnoLoggingFilter.doFilter(WebAnnoLoggingFilter.java:52)
     at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:235)
     at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:206)
     at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:330)
     at org.springframework.security.web.access.intercept.FilterSecurityInterceptor.invoke(FilterSecurityInterceptor.java:118)
     at org.springframework.security.web.access.intercept.FilterSecurityInterceptor.doFilter(FilterSecurityInterceptor.java:84)
     at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:342)
     at org.springframework.security.web.access.ExceptionTranslationFilter.doFilter(ExceptionTranslationFilter.java:113)
     at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:342)
     at org.springframework.security.web.session.SessionManagementFilter.doFilter(SessionManagementFilter.java:103)
     at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:342)
     at org.springframework.security.web.authentication.AnonymousAuthenticationFilter.doFilter(AnonymousAuthenticationFilter.java:113)
     at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:342)
     at org.springframework.security.web.authentication.rememberme.RememberMeAuthenticationFilter.doFilter(RememberMeAuthenticationFilter.java:146)
     at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:342)
     at org.springframework.security.web.servletapi.SecurityContextHolderAwareRequestFilter.doFilter(SecurityContextHolderAwareRequestFilter.java:54)
     at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:342)
     at org.springframework.security.web.savedrequest.RequestCacheAwareFilter.doFilter(RequestCacheAwareFilter.java:45)
     at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:342)
     at org.springframework.security.web.authentication.www.BasicAuthenticationFilter.doFilter(BasicAuthenticationFilter.java:150)
     at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:342)
     at org.springframework.security.web.authentication.AbstractAuthenticationProcessingFilter.doFilter(AbstractAuthenticationProcessingFilter.java:183)
     at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:342)
     at org.springframework.security.web.authentication.logout.LogoutFilter.doFilter(LogoutFilter.java:105)
     at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:342)
     at org.springframework.security.web.context.SecurityContextPersistenceFilter.doFilter(SecurityContextPersistenceFilter.java:87)
     at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:342)
     at org.springframework.security.web.FilterChainProxy.doFilterInternal(FilterChainProxy.java:192)
     at org.springframework.security.web.FilterChainProxy.doFilter(FilterChainProxy.java:160)
     at org.springframework.web.filter.DelegatingFilterProxy.invokeDelegate(DelegatingFilterProxy.java:346)
     at org.springframework.web.filter.DelegatingFilterProxy.doFilter(DelegatingFilterProxy.java:259)
     at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:235)
     at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:206)
     at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:233)
     at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:191)
     at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:127)
     at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:102)
     at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:109)
     at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:293)
     at org.apache.jk.server.JkCoyoteHandler.invoke(JkCoyoteHandler.java:190)
     at org.apache.jk.common.HandlerRequest.invoke(HandlerRequest.java:291)
     at org.apache.jk.common.ChannelSocket.invoke(ChannelSocket.java:776)
     at org.apache.jk.common.ChannelSocket.processConnection(ChannelSocket.java:705)
     at org.apache.jk.common.ChannelSocket$SocketConnection.runIt(ChannelSocket.java:898)
     at org.apache.tomcat.util.threads.ThreadPool$ControlRunnable.run(ThreadPool.java:690)
     at java.lang.Thread.run(Thread.java:745)

Complete stack:

org.apache.wicket.WicketRuntimeException: Method onFormSubmitted of interface org.apache.wicket.markup.html.form.IFormSubmitListener targeted at [ProjectDetailForm [Component id = projectDetailForm]] on component [ProjectDetailForm [Component id = projectDetailForm]] threw an exception
     at org.apache.wicket.RequestListenerInterface.internalInvoke(RequestListenerInterface.java:268)
     at org.apache.wicket.RequestListenerInterface.invoke(RequestListenerInterface.java:216)
     at org.apache.wicket.core.request.handler.ListenerInterfaceRequestHandler.invokeListener(ListenerInterfaceRequestHandler.java:243)
     at org.apache.wicket.core.request.handler.ListenerInterfaceRequestHandler.respond(ListenerInterfaceRequestHandler.java:236)
     at org.apache.wicket.request.cycle.RequestCycle$HandlerExecutor.respond(RequestCycle.java:890)
     at org.apache.wicket.request.RequestHandlerStack.execute(RequestHandlerStack.java:64)
     at org.apache.wicket.request.cycle.RequestCycle.execute(RequestCycle.java:261)
     at org.apache.wicket.request.cycle.RequestCycle.processRequest(RequestCycle.java:218)
     at org.apache.wicket.request.cycle.RequestCycle.processRequestAndDetach(RequestCycle.java:289)
     at org.apache.wicket.protocol.http.WicketFilter.processRequestCycle(WicketFilter.java:259)

java.lang.reflect.InvocationTargetException
     at sun.reflect.GeneratedMethodAccessor215.invoke(Unknown Source)
     at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
     at java.lang.reflect.Method.invoke(Method.java:497)
     at org.apache.wicket.RequestListenerInterface.internalInvoke(RequestListenerInterface.java:258)
     at org.apache.wicket.RequestListenerInterface.invoke(RequestListenerInterface.java:216)
     at org.apache.wicket.core.request.handler.ListenerInterfaceRequestHandler.invokeListener(ListenerInterfaceRequestHandler.java:243)
     at org.apache.wicket.core.request.handler.ListenerInterfaceRequestHandler.respond(ListenerInterfaceRequestHandler.java:236)
     at org.apache.wicket.request.cycle.RequestCycle$HandlerExecutor.respond(RequestCycle.java:890)
     at org.apache.wicket.request.RequestHandlerStack.execute(RequestHandlerStack.java:64)
     at org.apache.wicket.request.cycle.RequestCycle.execute(RequestCycle.java:261)
     at org.apache.wicket.request.cycle.RequestCycle.processRequest(RequestCycle.java:218)
     at org.apache.wicket.request.cycle.RequestCycle.processRequestAndDetach(RequestCycle.java:289)
     at org.apache.wicket.protocol.http.WicketFilter.processRequestCycle(WicketFilter.java:259)
```

**Comment by aakashysharma** [comment]
If encoding of file is ANSI, it works without error.

**Comment by aakashysharma** [comment]
Opened enhancement #96

**Comment by aakashysharma** [comment]
A UTF-8 encoded file works fine and is exported as UTF-8 from WebAnno.
If the same file, without modification is imported, it works.

It happens when you "Save As" an exported constraints file in UTF-8 using Notepad on Windows OS or modify and save it. 

I haven't tested other OS.

**Comment by aakashysharma** [comment]
The issue is in Windows 10 Notepad application. It's adding a special character at the beginning when I save file in UTF-8.

See the 4th line (I saved copy of same file which was already in UTF-8 without any modification), it adds a weird UTF-8 character which caused rules to fail when I tried uploading modified modified file. In ANSI encoding, it doesn't add that but non-English characters become "diamonds" �.

As far as webanno constraints module is concerned, it maintains the encoding (whichever [ANSI/UTF-8] it was at the time of upload) during import/export.

![issue 95](https://cloud.githubusercontent.com/assets/1897100/9670475/0d45dd9a-528d-11e5-81c0-859e1d51688a.png)
